apps/web/pages/api/twilio/webhook.ts

```diff
@@ -0,0 +1,164 @@
+import type { NextApiRequest, NextApiResponse } from "next";
+import { z } from "zod";
+
+import { CreditService } from "@calcom/features/ee/billing/credit-service";
+import * as twilio from "@calcom/features/ee/workflows/lib/reminders/providers/twilioProvider";
+import { IS_SMS_CREDITS_ENABLED, WEBAPP_URL } from "@calcom/lib/constants";
+import getOrgIdFromMemberOrTeamId from "@calcom/lib/getOrgIdFromMemberOrTeamId";
+import { defaultHandler } from "@calcom/lib/server";
+import prisma from "@calcom/prisma";
+
+const InputSchema = z.object({
+  userId: z
+    .string()
+    .optional()
+    .transform((val) => {
+      const num = Number(val);
+      return isNaN(num) ? undefined : num;
+    }),
+  teamId: z
+    .string()
+    .optional()
+    .transform((val) => {
+      const num = Number(val);
+      return isNaN(num) ? undefined : num;
+    }),
+  bookingUid: z.string().optional(),
+});
+
+/*
+  Twilio status callback: creates expense log when sms is delivered or undelivered
+*/
+async function handler(req: NextApiRequest, res: NextApiResponse) {
+  const signature = req.headers["x-twilio-signature"];
+  const baseUrl = `${WEBAPP_URL}/api/twilio/webhook`;
+
+  const queryParams = new URLSearchParams(req.query as Record<string, string>).toString();
+  const requestUrl = queryParams ? `${baseUrl}?${queryParams}` : baseUrl;
+
+  if (typeof signature !== "string") {
+    return res.status(401).send("Missing Twilio signature");
+  }
+
+  const isSignatureValid = await twilio.validateWebhookRequest({
+    requestUrl,
+    signature,
+    params: req.body,
+  });
+
+  if (!isSignatureValid) {
+    return res.status(401).send("Invalid Twilio signature");
+  }
+
+  const messageStatus = req.body.MessageStatus;
+
+  if (messageStatus !== "delivered" && messageStatus !== "undelivered") {
+    return res.status(200).send(`SMS not yet delivered/undelivered`);
+  }
+
+  if (!IS_SMS_CREDITS_ENABLED) {
+    return res.status(200).send(`SMS credits are not enabled.`);
+  }
+
+  const countryCode = await twilio.getCountryCodeForNumber(req.body.To);
+
+  const smsSid = req.body.SmsSid;
+
+  const {
+    userId: parsedUserId,
+    teamId: parsedTeamId,
+    bookingUid: parsedBookingUid,
+  } = InputSchema.parse(req.query);
```

Comment:
Much better! In addition, I think we can use safeParse and handle the error explicitly.